File: benchmark/primitive_ops/canisters/azle/data_types/int/int64.ts
import { ic, int64, nat32, nat64, Update } from 'azle';

let int64_init_heap_storage: { [key: string]: int64 | undefined; } = {};

export function int64_init_stack(num_inits: nat32): Update<nat64> {
    let i = 0;

    while (i < num_inits) {
        let value: int64 = i % 2 === 0 ? 9_223_372_036_854_775_807n : 0n;
        // TODO std::convert::identity(value); consider something like Rust to ensure the value assignment above is never optimized away
        i += 1;
    }

    return ic.performance_counter(0);
}

export function int64_init_heap(num_inits: nat32): Update<nat64> {
    let i = 0;

    while (i < num_inits) {
        int64_init_heap_storage[`element${i}`] = i % 2 === 0 ? 9_223_372_036_854_775_807n : 0n;
        i += 1;
    }

    return ic.performance_counter(0);
}

File: benchmark/primitive_ops/canisters/rust/src/data_types/int/int64.rs
use std::collections::HashMap;
use std::cell::RefCell;

thread_local! {
    static INT64_INIT_HEAP_STORAGE_REF_CELL: RefCell<Int64InitHeapStorage> = RefCell::default();
}

type Int64InitHeapStorage = HashMap<String, i64>;

#[ic_cdk_macros::update]
pub fn int64_init_stack(num_inits: u32) -> u64 {
    let mut i = 0;

    while i < num_inits {
        let value: i64 = if i % 2 == 0 { 9_223_372_036_854_775_807 } else { 0 };
        bencher::black_box(&value); // Trying to ensure that the value assignment above is not optimized away
        i += 1;
    }

    ic_cdk::api::call::performance_counter(0)
}

#[ic_cdk_macros::update]
pub fn int64_init_heap(num_inits: u32) -> u64 {
    INT64_INIT_HEAP_STORAGE_REF_CELL.with(|int64_init_heap_storage_ref_cell| {
        let mut i = 0;
        let mut int64_init_heap_storage = int64_init_heap_storage_ref_cell.borrow_mut();

        while i < num_inits {
            int64_init_heap_storage.insert(
                format!("element{}", i),
                if i % 2 == 0 { 9_223_372_036_854_775_807 } else { 0 }
            );
            
            i += 1;
        }
    });

    ic_cdk::api::call::performance_counter(0)
}